src/main/java/oracle/kubernetes/operator/wlsconfig/WlsServerConfig.java

```diff
@@ -79,4 +88,159 @@ public String toString() {
+          // The targets from JSON result looks like: [{identity=[servers, admin-server]}, {identity=[clusters, cluster-1]},
+          //                                           {identity=[servers, ms3]}, {identity=[clusters,cluster-2]}]
+          // Change this targets Map to Map<String, List<String>> type for easy handling later:
+          // e.g., [servers={admin-server,ms3}, clusters={cluster-1,cluster-2}]
+
+          // targets looks like this: [clusters={cluster-1, cluster-2},servers={ms3,ms4}]
+          Map<String, List<String>> targets = new HashMap<String, List<String>>();
+
+          List<String> serversList = new ArrayList<String>();
+          List<String> clustersList = new ArrayList<String>();
+
+          List<Map<String, Object>> targetsList = (List<Map<String, Object>>)thisApp.get("targets");
+
+          if (targetsList != null) {
+            for (Map<String, Object> thisTargets : targetsList) {
+              List<String> targetIdentity = (List<String>) thisTargets.get("identity");
+
+
+              // targetIdentity = [servers, admin-server]
+              // convert each targetIdentity list to a MapEntry: servers=admin-server
+              if (targetIdentity != null) {
+
+                String firstElem = targetIdentity.get(0);
+
+                if (firstElem.equals("servers")) {
+                  serversList.add(targetIdentity.get(1));
+                } else if (firstElem.equals("clusters")) {
+                  clustersList.add(targetIdentity.get(1));
+                }
+              }
+            }
+          }
+          targets.put("servers", serversList);
+          targets.put("clusters", clustersList);
+
+          wlsAppConfig.setTargets(targets);
+
+          LOGGER.fine(wlsAppConfig.toString());
+
+          // add to the map
+          wlsAppConfigMap.put(appNameFromJson, wlsAppConfig);
+
+        }
+
+        return wlsAppConfigMap;
+      }
+
+    } catch (Exception e) {
+      LOGGER.warning(MessageKeys.JSON_PARSING_FAILED, appJsonResult, e.getMessage());
+      throw e;
+    }
+    return null;
+
+  }
+
+  private static boolean needUpgrade(UpgradeApplicationsModel appsToPatch, String appName) {
+    List<UpgradeApplicationModel> appList = appsToPatch.getApplications();
+    for (UpgradeApplicationModel theApp : appList) {
+      if (theApp.getApplicationName().equals(appName)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private static String getPatchedLocation(UpgradeApplicationsModel appsToPatch, String appName) {
+    List<UpgradeApplicationModel> appList = appsToPatch.getApplications();
+    String patchedLoc = null;
+    for (UpgradeApplicationModel theApp : appList) {
+      if (theApp.getApplicationName().equals(appName)) {
+        patchedLoc = theApp.getPatchedLocation();
+        if (patchedLoc != null) {
+          if (patchedLoc.isEmpty()) {
+            throw new IllegalArgumentException(MessageKeys.APPUPGRADE_MISSING_PATCHED_LOCATION + " " + appName);
```

Comment:
This looks odd.  What are you doing here?

Reply:
this is trying to verify if customer provides incomplete info for app upgrade when posting the upgrade operation, such as, no patchedLocation, no backupLocation, we will error this out and no further upgrade processing.